internal/pkg/rule/cel.go

```diff
@@ -0,0 +1,291 @@
+	constExpr, ok := expression.ExprKind.(*expr.Expr_ConstExpr)
+	if !ok {
+		return function.OrderTypeUnknown, fmt.Errorf("sequence function second argument requires a constant string with value 'asc' or 'desc'")
+	}
+
+	stringExpr, ok := constExpr.ConstExpr.ConstantKind.(*expr.Constant_StringValue)
+	if !ok {
+		return function.OrderTypeUnknown, fmt.Errorf("sequence function second argument requires a constant string with value 'asc' or 'desc'")
+	}
+
+	var order function.OrderType
+	switch stringExpr.StringValue {
+	case "asc":
+		order = function.OrderTypeAsc
+	case "desc":
+		order = function.OrderTypeDesc
+	default:
+		return function.OrderTypeUnknown, fmt.Errorf("sequence function second argument requires a constant string with value 'asc' or 'desc'")
+	}
+
+	return order, nil
+}
+
+type CheckedExprModifier struct {
+	currentStateID int64
+	currentExprID  int64
+	CheckedExpr    *expr.CheckedExpr
+}
+
+func NewCheckedExprModifier(checkedExpr *expr.CheckedExpr) *CheckedExprModifier {
+	id := int64(0)
+	for i := range checkedExpr.ReferenceMap {
+		if i > id {
+			id = i
+		}
+	}
+	id++
+
+	return &CheckedExprModifier{
+		currentStateID: 0,
+		currentExprID:  id,
+		CheckedExpr:    checkedExpr,
+	}
+}
+
+func (cem *CheckedExprModifier) injectStateToCall(exprID int64, callExpr *expr.Expr_Call, messageType string) string {
+	stateName := "state" + fmt.Sprintf("%d", cem.currentStateID)
+
+	argumentExpression := &expr.Expr{
+		Id: cem.currentExprID,
+		ExprKind: &expr.Expr_IdentExpr{
+			IdentExpr: &expr.Expr_Ident{
+				Name: stateName,
+			},
+		},
+	}
+
+	// Add state argument
+	callExpr.Args = append(callExpr.Args, argumentExpression)
+
+	// Add expression id to type map
+	cem.CheckedExpr.TypeMap[cem.currentExprID] = &expr.Type{
+		TypeKind: &expr.Type_MessageType{
+			MessageType: messageType,
+		},
+	}
+
+	// Modify overloads in the reference map
+	for i, overload := range cem.CheckedExpr.ReferenceMap[exprID].OverloadId {
+		cem.CheckedExpr.ReferenceMap[exprID].OverloadId[i] = overload + "_state"
+	}
+
+	cem.currentStateID++
+	cem.currentExprID++
+
+	return stateName
+}
+
+func (cem *CheckedExprModifier) InjectState() ([]*function.StatefulFunctionProvider, error) {
+	return cem.injectState(cem.CheckedExpr.Expr)
+}
+
+func (cem *CheckedExprModifier) injectState(expression *expr.Expr) ([]*function.StatefulFunctionProvider, error) {
+	if expression == nil {
+		return []*function.StatefulFunctionProvider{}, nil
+	}
+
+	providers := []*function.StatefulFunctionProvider{}
+	switch expression.ExprKind.(type) {
+	case *expr.Expr_SelectExpr:
+		selectExpr := expression.GetSelectExpr()
+
+		// Recursive check of the operand
+		selectStatefulFunctions, err := cem.injectState(selectExpr.Operand)
+		if err != nil {
+			return []*function.StatefulFunctionProvider{}, err
+		}
+		providers = append(providers, selectStatefulFunctions...)
+
+	case *expr.Expr_CallExpr:
+		callExpr := expression.GetCallExpr()
+
+		switch callExpr.Function {
```

Comment:
is it safe to not have a `default`?

Reply:
yes, in fact is needed. When a function is a stateless function we don't have to do anything with the state, but it's still a valid case.